payments/dotpay/forms.py

```diff
@@ -43,34 +46,48 @@ def clean(self):
+                cleaned_data.get('id', ''),
+                cleaned_data.get('operation_number', ''),
+                cleaned_data.get('operation_type', ''),
+                cleaned_data.get('operation_status', ''),
+                cleaned_data.get('operation_amount', ''),
+                cleaned_data.get('operation_currency', ''),
+                cleaned_data.get('operation_withdrawal_amount', ''),
+                cleaned_data.get('operation_commission_amount', ''),
+                cleaned_data.get('is_completed', ''),
+                cleaned_data.get('operation_original_amount', ''),
+                cleaned_data.get('operation_original_currency', ''),
+                cleaned_data.get('operation_datetime', ''),
+                cleaned_data.get('operation_related_number', ''),
+                cleaned_data.get('control', ''),
+                cleaned_data.get('description', ''),
                 cleaned_data.get('email', ''),
-                '',  # service
-                '',  # code
-                '',  # username
-                '',  # password
-                str(cleaned_data['t_status']))
-            key = ':'.join(key_vars)
-            md5 = hashlib.md5()
-            md5.update(key.encode('utf-8'))
-            key_hash = md5.hexdigest()
-            if key_hash != self.cleaned_data['md5']:
-                self._errors['md5'] = self.error_class(['Bad hash'])
-            if cleaned_data['control'] != self.payment.id:
+                cleaned_data.get('p_info', ''),
+                cleaned_data.get('p_email', ''),
+                cleaned_data.get('credit_card_issuer_identification_number', ''),
+                cleaned_data.get('credit_card_masked_number', ''),
+                cleaned_data.get('credit_card_brand_codename', ''),
+                cleaned_data.get('credit_card_brand_code', ''),
+                cleaned_data.get('credit_card_id', ''),
+                cleaned_data.get('channel', ''),
+                cleaned_data.get('channel_country', ''),
+                cleaned_data.get('geoip_country', ''))
+            key = ''.join(key_vars)
+            sha256 = hashlib.sha256()
+            sha256.update(key.encode('utf-8'))
+            key_hash = sha256.hexdigest()
+            if key_hash != self.cleaned_data['signature']:
+                self._errors['signature'] = self.error_class(['Bad hash'])
+            if int(cleaned_data['control']) != self.payment.id:
                 self._errors['control'] = self.error_class(['Bad payment id'])
         return cleaned_data
 
     def save(self, *args, **kwargs):
-        status = self.cleaned_data['t_status']
-        self.payment.transaction_id = self.cleaned_data['t_id']
+        status = self.cleaned_data['operation_status']
+        self.payment.transaction_id = self.cleaned_data['operation_number']
         self.payment.save()
-        payment_status = self.payment.status
-        if status == ACCEPTED:
+        if status == 'completed':
```

Comment:
I think it's a good idea to keep such statuses in constants - it improves readability and allows easier reusability for the future. I'm not sure if all of those that you removed at the top of the original file were used and correctly named. Something like that may work:

```
COMPLETED = 'completed'
REJECTED = 'REJECTED'
```

Reply:
Ok, will map statuses like you suggest. The statuses I removed are not mentioned in the new docs released by Dotpay. Currently there are only the following statuses:

new
processing
completed
rejected
processing_realization_waiting
processing_realization

of which of interest are "processing" and "completed" (at least in our context here). Ignoring the other statuses should not do any harm, I don't think they will ever be used in this form. In any case, we save the transaction number for reference.